spring-integration-test/src/main/java/org/springframework/integration/test/mock/MockMessageHandler.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.test.mock;
+
+import static org.junit.Assert.assertThat;
+
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+import org.hamcrest.Matcher;
+
+import org.springframework.integration.handler.AbstractReplyProducingMessageHandler;
+import org.springframework.messaging.Message;
+
+/**
+ * @author Artem Bilan
+ *
+ * @since 5.0
+ */
+public class MockMessageHandler extends AbstractReplyProducingMessageHandler {
```

Comment:
The idea behind this component is excellent but the API is quite hard to understand (for me of course). We know that the readability is one of the first requirements to test code so I'd like to pay your attention to a couple of points to improve. In particular:

- it is not clear at the first glance that the `expect`/[`andReply`] methods chain composes a kind of one-time-use scenario for the handler mock. It may seem that they just describe different variations of conditions for unlimited matchings against incoming messages. To avoid misuse the purpose of methods should be clearly stated in JavaDoc at least;
- `andReply` methods name may confuse end user in case of trying to read the chain in a straightforward manner, for instance, the fragment `.expect(hasHeader("baz", "BAZ")).andReply()` may seem like expecting a reply (like in `FileWritingMessageHandler#setExpectReply` method). I think it would be better to rename it to `thenReply` with some additions (see further);
- no-arg version of `andReply()` method is not self-describable as it is not obvious that the method involves the `Function#identity` function. Perhaps, it would be better to rename to something like `thenReplyWithRequest`.